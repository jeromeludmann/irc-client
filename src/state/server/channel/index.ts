import { combineReducers } from "redux";
import { createSelector } from "reselect";
import input, { InputState } from "@app/state/server/channel/input";
import messages, {
  MessageListState,
  getMessages,
} from "@app/state/server/channel/message-list";

export interface ChannelState {
  messages: MessageListState;
  input: InputState;
}

export default combineReducers<ChannelState>({
  messages,
  input,
});

export const getMessagesCount = createSelector(
  getMessages,
  msg => (msg ? msg.length : 0),
);
