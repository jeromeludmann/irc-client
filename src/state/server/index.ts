import { Action } from "redux";
import reduceModes, {
  ModesState,
  modesInitialState,
} from "@app/state/server/modes";
import reduceChannels, {
  ChannelRouterState,
  ChannelRouterAction,
  channelsInitialState,
} from "@app/state/server/channel-router";
import { Route } from "@app/state/Route";
import { ActiveRouteState } from "@app/state/active";

export type ServerState = {
  readonly modes: ModesState;
  readonly channels: ChannelRouterState;
};

export type ServerAction = Action;

interface ExtraParams {
  route: Route;
  active: ActiveRouteState;
}
export const serverInitialState: ServerState = {
  channels: channelsInitialState,
  modes: modesInitialState,
};

export default function reduceServer(
  server = serverInitialState,
  action: ServerAction,
  { route, active }: ExtraParams,
): ServerState {
  return {
    modes: reduceModes(server.modes),
    channels: reduceChannels(server.channels, action as ChannelRouterAction, {
      route,
      active,
    }),
  };
}
