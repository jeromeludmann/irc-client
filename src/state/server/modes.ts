export type ModesState = string[];

export const modesInitialState: ModesState = [];

export default (modes = modesInitialState): ModesState => modes;
