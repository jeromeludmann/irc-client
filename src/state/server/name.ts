export type NameState = string;

type NameAction = {
  type: "SET_NAME";
  payload: { name: string };
};

const initialState: NameState = "";

export default function(name = initialState, action: NameAction): NameState {
  switch (action.type) {
    case "SET_NAME":
      return action.payload.name;
    default:
      return name;
  }
}
