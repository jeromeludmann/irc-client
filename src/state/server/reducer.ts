import { AnyAction, Action } from 'redux'
import { RouteState } from '@app/state/route/reducer'
import {
  RoutedAction,
  BufferKey,
  isChannel,
  isPrivate,
  isRaw,
} from '@app/utils/Route'
import {
  RECEIVE_RPL_MYINFO,
  ReceiveReplyMyInfoAction,
  RECEIVE_NICK,
  ReceiveNickAction,
  RECEIVE_PART,
  ReceivePartAction,
  RECEIVE_PRIVMSG,
  RECEIVE_JOIN,
} from '@app/actions/messages/incoming'
import {
  BufferState,
  bufferInitialState,
  reduceBuffer,
} from '@app/state/buffer/reducer'
import {
  CLOSE_WINDOW,
  CloseWindowAction,
  UPDATE_SERVER_LAG,
  UpdateServerLagAction,
} from '@app/actions/ui'

export type ServerState = Readonly<{
  name: string
  user: Readonly<{
    nick: string
    user: string
    real: string
  }>
  lag: number
  modes: Readonly<{
    user: string[]
    available: Readonly<{
      channel: string[]
      user: string[]
    }>
  }>
  buffers: Readonly<{
    [key: string]: BufferState
  }>
}>

type ServerReducer<S = ServerState> = (
  server: S,
  action: Action,
  extraStates: { route: RouteState },
) => S

type BufferRouterReducer = (
  buffers: { [key: string]: BufferState },
  action: RoutedAction,
  extraStates: { route: RouteState; server: ServerState },
) => { [key: string]: BufferState }

export const serverInitialState = {
  name: '<unknown>',
  user: {
    nick: 'default_nick',
    user: 'default_user',
    real: 'default_name',
  },
  lag: 0,
  modes: {
    user: [],
    available: { channel: [], user: [] },
  },
  buffers: {
    [BufferKey.RAW]: bufferInitialState,
    [BufferKey.STATUS]: bufferInitialState,
  },
}

const removeCurrentBuffer = (
  buffers: { [key: string]: BufferState },
  bufferKey: string,
) => {
  const bufferMap = { ...buffers }
  delete bufferMap[bufferKey]
  return bufferMap
}

const removeAllServerRelatedBuffers = (buffers: {
  [key: string]: BufferState
}) => {
  const bufferMap = { ...buffers }

  Object.keys(bufferMap).forEach(buffer => {
    if (isChannel(buffer) || isPrivate(buffer)) {
      delete bufferMap[buffer]
    }
  })

  return bufferMap
}

const itIsMe = (server: ServerState, action: AnyAction) =>
  action.payload.user.nick === server.user.nick

const caseReducers: { [action: string]: ServerReducer } = {
  [CLOSE_WINDOW]: (server, action: CloseWindowAction) => ({
    ...server,
    buffers:
      isChannel(action.route.bufferKey) || isPrivate(action.route.bufferKey)
        ? removeCurrentBuffer(server.buffers, action.route.bufferKey)
        : removeAllServerRelatedBuffers(server.buffers),
  }),

  [RECEIVE_NICK]: (server, action: ReceiveNickAction) => ({
    ...server,
    user: itIsMe(server, action)
      ? { ...server.user, nick: action.payload.nick }
      : server.user,
  }),

  // We arbitrarily decided to close window when we "/part" the channel.
  // But later, we could make this behavior customizable.
  [RECEIVE_PART]: (server, action: ReceivePartAction) =>
    itIsMe(server, action)
      ? {
          ...server,
          buffers: removeCurrentBuffer(server.buffers, action.payload.channel),
        }
      : server,

  [RECEIVE_RPL_MYINFO]: (server, action: ReceiveReplyMyInfoAction) => ({
    ...server,
    name: action.payload.serverName,
    modes: {
      ...server.modes,
      available: {
        channel: action.payload.availableChannelModes,
        user: action.payload.availableUserModes,
      },
    },
  }),

  [UPDATE_SERVER_LAG]: (server, action: UpdateServerLagAction) => ({
    ...server,
    lag: action.payload.lag,
  }),
}

const broadcastHandlers: { [buffer: string]: BufferRouterReducer } = {
  // stop broadcasting
  [BufferKey.NONE]: buffers => buffers,

  // broadcast to active buffer
  [BufferKey.ACTIVE]: (buffers, action, extraStates) => {
    const key = extraStates.route.bufferKey
    return {
      ...buffers,
      [key]: reduceBuffer(buffers[key], action, extraStates),
    }
  },

  // broadcast to all buffers
  [BufferKey.ALL]: (buffers, action, extraStates) => {
    const toBroadcast: { [key: string]: BufferState } = {}

    Object.keys(buffers).forEach(key => {
      if (!isRaw(key)) {
        toBroadcast[key] = reduceBuffer(buffers[key], action, extraStates)
      }
    })

    return { ...buffers, ...toBroadcast }
  },
}

const routeActionToBuffers: BufferRouterReducer = (
  buffers,
  action,
  extraStates,
) => {
  if (action.route.bufferKey in broadcastHandlers) {
    return broadcastHandlers[action.route.bufferKey](
      buffers,
      action,
      extraStates,
    )
  }

  if (
    action.route.bufferKey in buffers ||
    [RECEIVE_PRIVMSG, RECEIVE_JOIN].indexOf(action.type) > -1
  ) {
    return {
      ...buffers,
      [action.route.bufferKey]: reduceBuffer(
        buffers[action.route.bufferKey],
        action,
        extraStates,
      ),
    }
  }

  return buffers
}

export const reduceServer: ServerReducer = (
  server = serverInitialState,
  action,
  extraStates,
) => {
  const intermediateState = {
    ...server,
    buffers: routeActionToBuffers(server.buffers, action as RoutedAction, {
      ...extraStates,
      server,
    }),
  }

  return {
    ...intermediateState,
    ...(action.type in caseReducers
      ? caseReducers[action.type](intermediateState, action, extraStates)
      : {}),
  }
}
