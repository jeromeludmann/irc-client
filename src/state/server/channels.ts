import channel, { ChannelState } from "@app/state/channel";
import { RoutedAction, ChannelRoute } from "@app/Route";

export interface ChannelListState {
  [key: string]: ChannelState;
}

const initialState = {
  status: {
    messages: [],
    input: { value: "", history: [] },
    unread: false,
  },
};

export default function(
  channels: ChannelListState = initialState,
  action: RoutedAction<ChannelRoute>,
): ChannelListState {
  return action.route && action.route.channel
    ? {
        ...channels,
        [action.route.channel]: channel(channels[action.route.channel], action),
      }
    : channels;
}
