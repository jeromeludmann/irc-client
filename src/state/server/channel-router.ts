import reduceChannel, { ChannelState } from "@app/state/channel";
import { RoutedAction, Route } from "@app/actions/Route";
import { ActiveState } from "@app/state/active";

export interface ChannelRouterState {
  readonly [key: string]: ChannelState;
}

export type ChannelRouterAction = RoutedAction;

interface ExtraParams {
  readonly route: Route;
  readonly active: ActiveState;
}

export const channelsInitialState: ChannelRouterState = {};

export default function(
  channels = channelsInitialState,
  action: ChannelRouterAction,
  { route, active }: ExtraParams,
): ChannelRouterState {
  return route.channel
    ? {
        ...channels,
        [route.channel]: reduceChannel(channels[route.channel], action, {
          active,
        }),
      }
    : channels;
}
