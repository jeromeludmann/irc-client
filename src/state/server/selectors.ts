import { createSelector } from "reselect";
import { selectServer, selectActiveRoute } from "@app/state/selectors";

export const selectChannel = createSelector(
  selectServer,
  selectActiveRoute,
  (server, active) => server.channels[active.channel],
);
