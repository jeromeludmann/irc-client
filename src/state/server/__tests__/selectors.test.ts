import { getServer, getBuffers } from '@app/state/server/selectors'
import { rootInitialState } from '@app/state/root/reducer'

describe('select server state', () => {
  it('should select server', () => {
    expect(getServer()(rootInitialState)).toMatchSnapshot()
  })

  it('should select buffers', () => {
    expect(getBuffers()(rootInitialState)).toMatchSnapshot()
  })
})
