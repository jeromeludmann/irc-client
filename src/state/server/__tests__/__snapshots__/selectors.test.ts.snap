// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`select server state should select buffers 1`] = `
Object {
  "@raw": Object {
    "activity": false,
    "input": Object {
      "dirtyValue": "",
      "history": Object {
        "index": 0,
        "values": Array [],
      },
      "value": "",
    },
    "messages": Array [],
  },
  "@status": Object {
    "activity": false,
    "input": Object {
      "dirtyValue": "",
      "history": Object {
        "index": 0,
        "values": Array [],
      },
      "value": "",
    },
    "messages": Array [],
  },
}
`;

exports[`select server state should select server 1`] = `
Object {
  "buffers": Object {
    "@raw": Object {
      "activity": false,
      "input": Object {
        "dirtyValue": "",
        "history": Object {
          "index": 0,
          "values": Array [],
        },
        "value": "",
      },
      "messages": Array [],
    },
    "@status": Object {
      "activity": false,
      "input": Object {
        "dirtyValue": "",
        "history": Object {
          "index": 0,
          "values": Array [],
        },
        "value": "",
      },
      "messages": Array [],
    },
  },
  "lag": 0,
  "modes": Object {
    "available": Object {
      "channel": Array [],
      "user": Array [],
    },
    "user": Array [],
  },
  "name": "<unknown>",
  "user": Object {
    "nick": "default_nick",
    "real": "default_name",
    "user": "default_user",
  },
}
`;
